datumaro/components/extractor.py

```diff
@@ -170,22 +178,97 @@ def get(self, id, subset=None):
                 return item
         return None
 
+T = TypeVar('T')
+
+class _ImportFail(DatumaroError):
+    pass
+
+class ImportErrorPolicy:
+    def report_item_error(self, error: Exception, *,
+            item_id: Tuple[str, str]):
+        """
+        Allows to report a problem with a dataset item.
+        If this function returns, the extractor must skip the item.
+        """
+
+        if not isinstance(error, _ImportFail):
+            ie = ItemImportError(item_id)
+            ie.__cause__ = error
+            return self._handle_item_error(ie)
+        else:
+            raise error
+
+    def report_annotation_error(self, error: Exception, *,
+            item_id: Tuple[str, str]):
+        """
+        Allows to report a problem with a dataset item annotation.
+        If this function returns, the extractor must skip the item.
+        """
+
+        if not isinstance(error, _ImportFail):
+            ie = AnnotationImportError(item_id)
+            ie.__cause__ = error
+            return self._handle_annotation_error(ie)
+        else:
+            raise error
+
+    def _handle_item_error(self, error: ItemImportError):
+        """This function must either call fail() or return."""
+        raise NotImplementedError("Must be defined in the subclass")
+
+    def _handle_annotation_error(self, error: AnnotationImportError):
+        """This function must either call fail() or return."""
+        raise NotImplementedError("Must be defined in the subclass")
+
+    def fail(self, error: Exception) -> NoReturn:
+        raise _ImportFail from error
+
+class FailingImportErrorPolicy(ImportErrorPolicy):
+    def _handle_item_error(self, error: ItemImportError):
+        self.fail(error)
+
+    def _handle_annotation_error(self, error: AnnotationImportError):
+        self.fail(error)
+
+@define(eq=False)
+class ImportContext:
+    progress_reporter: ProgressReporter = field(default=None,
+        converter=attr.converters.default_if_none(factory=NullProgressReporter))
+    error_policy: ImportErrorPolicy = field(default=None,
+        converter=attr.converters.default_if_none(factory=FailingImportErrorPolicy))
+
+class NullImportContext(ImportContext):
```

Comment:
This class seems useless. It's only used once and in that instance it could just be replaced by `ImportContext`.

Reply:
From my point, it clearly designates the caller intention, while provision of defaults in the base class is more of a forward compatibility.

Reply:
I don't really understand the nuance here. Is there any circumstance where `ImportContext()` would act any differently from `NullImportContext()`? You're just giving the user a pointless choice.

In addition, this enables users to write nonsense like `NullImportContext(error_policy=...)`, which seems undesirable.

Reply:
> I don't really understand the nuance here. Is there any circumstance where ImportContext() would act any differently from NullImportContext()? You're just giving the user a pointless choice.

I see it that way that ImportContext is just a base class. It doesn't have to be usable, though it can provide convenient defaults. 

> In addition, this enables users to write nonsense like NullImportContext(error_policy=...), which seems undesirable.

Why they would need to do this?